cobbler/utils/threading.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module is responsible for managing the custom common threading logic Cobbler has.
+"""
+
+import logging
+import pathlib
+import time
+from threading import Thread
+
+from cobbler import enums
+from cobbler import utils
+
+
+class CobblerThread(Thread):
+    """
+    This is a custom thread that has a custom logger as well as logic to execute Cobbler triggers.
+    """
+
+    def __init__(self, event_id: str, remote, options: dict, task_name: str, api):
+        """
+        This constructor creates a Cobbler thread which then may be run by calling ``run()``.
+
+        :param event_id: The event-id which is associated with this thread. Also used as thread name
+        :param remote: The Cobbler remote object to execute actions with.
+        :param options: Additional options which can be passed into the Thread.
+        :param task_name: The high level task name which is used to trigger pre- and post-task triggers
+        :param api: The Cobbler api object to resolve information with.
+        """
+        Thread.__init__(self, name=event_id)
+        self.event_id = event_id
+        self.remote = remote
+        self.logger = logging.getLogger()
+        self.__setup_logger()
+        if options is None:
+            options = {}
+        self.options = options
+        self.task_name = task_name
+        self.api = api
+
+    def __setup_logger(self):
+        """
+        Utility function that will setup the Python logger for the tasks in a special directory.
+        """
+        filename = pathlib.Path("/var/log/cobbler/tasks") / f"{self.event_id}.log"
+        task_log_handler = logging.FileHandler(str(filename), encoding="utf-8")
+        task_log_formatter = logging.Formatter(
+            "[%(threadName)s] %(asctime)s - %(levelname)s | %(message)s"
+        )
+        task_log_handler.setFormatter(task_log_formatter)
+        self.logger.setLevel(logging.INFO)
+        self.logger.addHandler(task_log_handler)
+
+    def on_done(self):
+        """
+        This stub is needed to satisfy the Python inheritance chain.
```

Comment:
That sound weird, are you sure that it's needed and supposed to be empty?

Reply:
No idea. Need to deep dive.

Reply:
I will answer with this comment to all unresolved threads in regard to `on_done` and `_run`:

We have two "well-known" functions called `_run` and `on_done` that we assign functions to that are then executed. The assignment of the functions is done in `__start_task` in `remote.py`. I am not aware of a clean way to force Python to "know" about this concept.

In regard to `on_done`: The reason why this is existing is explained in the docstring of `__start_task` and we have a usage example in `background_buildiso` in `remote.py`.

In regard to `_run`: This is the function that will handle the real execution of the task. We are not using `run` since this is already used by our custom `CobblerThread` to write the logic that executes triggers and handles Exceptions.

Reply:
I'm not sure about the best way to handle this, but I think it would already be an improvement to pass the two functions (assigned to `_run` and `on_done`) to `CobblerThread.__init__`. Attaching these two functions from the outside makes it really hard to understand where they are coming from when looking at `CobblerThread`.

Reply:
So I tried applying your suggestions. Please have a look again. Many thanks for the inspiration.